### PR TITLE
Add aten::index and aten::index_put

### PR DESCRIPTION
**Pull request by AlienLiang23**
1. Implement index kernel
2. Implement index put deterministic kernel
3. Add aten::index and aten::index_put

### CODE REVIEW

**Comment by fengyuan14** [comment]
Hold on merge. I will add example case for validate deterministic code-path.

File: src/aten/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -3,18 +3,122 @@
...
+
+#include <ATen/XPUNativeFunctions.h>
 #include <aten/sycl/IndexingKernel.h>
-#include <comm/RegisterUtils.h>
-#include <torch/library.h>
+#include <comm/ReduceOpsUtils.h>
 
 namespace at {
 
+// XXX: Should reuse source in storck PyTorch when in-tree.
```
Comment (fengyuan14) [inline src/aten/TensorAdvancedIndexing.cpp:19]:
```suggestion
// TODO: Should reuse source in storck PyTorch when in-tree.
```

File: src/aten/TensorAdvancedIndexing.cpp [hunk 2]
```diff
@@ -3,18 +3,122 @@
...
+
+#include <ATen/XPUNativeFunctions.h>
 #include <aten/sycl/IndexingKernel.h>
-#include <comm/RegisterUtils.h>
-#include <torch/library.h>
+#include <comm/ReduceOpsUtils.h>
 
 namespace at {
 
+// TODO: Should reuse source in storck PyTorch when in-tree.
```
Comment (EikanWang) [inline src/aten/TensorAdvancedIndexing.cpp:19]:
```suggestion
// TODO: Should reuse source in stock PyTorch when in-tree.
```

**Review by EikanWang — APPROVED** [review]
Please check if the test cases have covered all the functions.

**Comment by fengyuan14** [comment]
> Please check if the test cases have covered all the functions.

Yes, I have checked. Totally about 50+ cases are tested in TestCommon. And I also add example case for deterministic implementation.

**Pull request merged**